File: apps/client-macos/Sources/TimeTrack/Tracking/UUIDv7.swift
import Foundation

/// PRD §7.5 — every record carries a client-minted UUIDv7 primary key so sync is
/// idempotent (the API upserts on it). Hand-rolled to avoid a SwiftPM dependency
/// (CLAUDE.md §2). Layout (RFC 9562): 48-bit ms timestamp · version 7 · 12 rand ·
/// variant 10 · 62 rand. `now`/`randomByte` are injectable for deterministic tests.
enum UUIDv7 {
    static func generate(
        now: Date = Date(),
        randomByte: () -> UInt8 = { UInt8.random(in: UInt8.min...UInt8.max) }
    ) -> String {
        var bytes = [UInt8](repeating: 0, count: 16)

        // 48-bit big-endian millisecond timestamp.
        let ms = UInt64((now.timeIntervalSince1970 * 1000).rounded(.down))
        bytes[0] = UInt8((ms >> 40) & 0xFF)
        bytes[1] = UInt8((ms >> 32) & 0xFF)
        bytes[2] = UInt8((ms >> 24) & 0xFF)
        bytes[3] = UInt8((ms >> 16) & 0xFF)
        bytes[4] = UInt8((ms >> 8) & 0xFF)
        bytes[5] = UInt8(ms & 0xFF)

        for i in 6..<16 { bytes[i] = randomByte() }
        bytes[6] = (bytes[6] & 0x0F) | 0x70 // version 7
        bytes[8] = (bytes[8] & 0x3F) | 0x80 // variant 10

        let hex = bytes.map { String(format: "%02x", $0) }.joined()
        func slice(_ lo: Int, _ hi: Int) -> Substring {
            let a = hex.index(hex.startIndex, offsetBy: lo)
            let b = hex.index(hex.startIndex, offsetBy: hi)
            return hex[a..<b]
        }
        return "\(slice(0, 8))-\(slice(8, 12))-\(slice(12, 16))-\(slice(16, 20))-\(slice(20, 32))"
    }
}

File: apps/client-windows/src/NiftyTimer/Tracking/UuidV7.cs
using System.Globalization;

namespace NiftyTimer.Tracking;

/// <summary>
/// PRD §7.5 — every record carries a client-minted UUIDv7 primary key so sync is idempotent
/// (the API upserts on it). Hand-rolled to avoid a dependency (CLAUDE.md §2). Layout
/// (RFC 9562): 48-bit big-endian ms timestamp · version 7 · 12 rand · variant 10 · 62 rand.
/// <paramref name="now"/> / <paramref name="randomByte"/> are injectable for deterministic tests.
/// </summary>
public static class UuidV7
{
    public static string Generate(DateTimeOffset now, Func<byte>? randomByte = null)
    {
        randomByte ??= DefaultRandomByte;

        var bytes = new byte[16];

        // 48-bit big-endian millisecond timestamp.
        var ms = (ulong)now.ToUnixTimeMilliseconds();
        bytes[0] = (byte)((ms >> 40) & 0xFF);
        bytes[1] = (byte)((ms >> 32) & 0xFF);
        bytes[2] = (byte)((ms >> 24) & 0xFF);
        bytes[3] = (byte)((ms >> 16) & 0xFF);
        bytes[4] = (byte)((ms >> 8) & 0xFF);
        bytes[5] = (byte)(ms & 0xFF);

        for (var i = 6; i < 16; i++)
        {
            bytes[i] = randomByte();
        }

        bytes[6] = (byte)((bytes[6] & 0x0F) | 0x70); // version 7
        bytes[8] = (byte)((bytes[8] & 0x3F) | 0x80); // variant 10

        var hex = Convert.ToHexString(bytes).ToLowerInvariant();
        return string.Create(
            36,
            hex,
            static (span, h) =>
            {
                h.AsSpan(0, 8).CopyTo(span);
                span[8] = '-';
                h.AsSpan(8, 4).CopyTo(span[9..]);
                span[13] = '-';
                h.AsSpan(12, 4).CopyTo(span[14..]);
                span[18] = '-';
                h.AsSpan(16, 4).CopyTo(span[19..]);
                span[23] = '-';
                h.AsSpan(20, 12).CopyTo(span[24..]);
            });
    }

    public static string Generate() => Generate(DateTimeOffset.UtcNow);

    private static byte DefaultRandomByte() =>
        (byte)System.Security.Cryptography.RandomNumberGenerator.GetInt32(0, 256);

    /// <summary>
    /// The wire format the API expects for instants: RFC 3339 UTC, no fractional seconds
    /// (matches Swift's <c>ISO8601DateFormatter</c> with <c>.withInternetDateTime</c>, which is
    /// what the shipped Mac client sends and what <c>z.iso.datetime()</c> accepts).
    /// </summary>
    public static string Iso(DateTimeOffset instant) =>
        instant.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);
}
